cpu/atmega_common/periph/i2c.c

```diff
@@ -0,0 +1,347 @@
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+};
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    /* TWI Bit Rate Register - division factor for the bit rate generator*/
+    int twibrr;
+
+    /* check if the line is valid */
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* calculate speed configuration */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            twibrr = ((CLOCK_CORECLOCK/100000)-16)/2;
+            break;
+
+        case I2C_SPEED_FAST:
+            twibrr = ((CLOCK_CORECLOCK/400000)-16)/2;
+            break;
+
+        default:
+            return -2;
+    }
+
+    /* set pull-up on SCL and SDA */
+    #if defined (CPU_ATMEGA2560) || defined (CPU_ATMEGA1281)
+        PORTD |= (1 << PORTD0) | (1 << PORTD1);
+    #endif
+    #ifdef CPU_ATMEGA328P
+        PORTC |= (1 << PORTC4) | (1 << PORTC5);
+    #endif
+
+    /* enable I2C clock */
+    i2c_poweron(dev);
+
+    /* disable device */
+    TWCR &= ~(1 << TWEN);
+    /* configure I2C clock */
+    TWBR = twibrr;
+    /* enable device */
+    TWCR |= (1 << TWEN);
+
+    return 0;
+}
+
+int i2c_acquire(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_lock(&locks[dev]);
+    return 0;
+}
+
+int i2c_release(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int i2c_read_byte(i2c_t dev, uint8_t address, void *data)
+{
+    return i2c_read_bytes(dev, address, data, 1);
+}
+
+int i2c_read_bytes(i2c_t dev, uint8_t address, void *data, int length)
+{
+    uint8_t *my_data = data;
+
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* send start condition and slave address */
+    if (_start(address, I2C_FLAG_READ) != 0)
+        return 0;
+
+    for (int i = 0; i < length; i++) {
+        /* Send NACK for last received byte */
+        if ((length-i) == 1)
+            TWCR = (1 << TWEN) | (1 << TWINT);
+        else
+            TWCR = (1 << TWEA) | (1 << TWEN) | (1 << TWINT);
+        DEBUG("Wait for byte %i\n", i+1);
+        /* Wait for TWINT Flag set. This indicates that DATA has been received.*/
+        while (!(TWCR & (1 << TWINT)))
+            {}
+        /* receive data byte */
+        my_data[i] = TWDR;
+        DEBUG("Byte %i received\n", i+1);
+    }
+
+    /* end transmission */
+    _stop();
+
+    return length;
+}
+
+int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, void *data)
+{
+    return i2c_read_regs(dev, address, reg, data, 1);
+}
+
+int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg, void *data, int length)
+{
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* start transmission and send slave address */
+    if (_start(address, I2C_FLAG_WRITE) != 0)
+        return 0;
+
+    /* send register address and wait for complete transfer to be finished*/
+    if (_write(&reg, 1) != 1) {
+        _stop();
+        return 0;
+    }
+
+    _stop();
```

Comment:
I know this is not very good documented, but most devices expect a 'repeated start condition' when reading/writing registers, so I think this `_stop` should not be here, right?